support flip and rot90 for complex dtype

kshitij12345 opened the pull request:
Closes #37698

kshitij12345 committed:
support flip and rot90 for complex dtype

kshitij12345 commented:
@anjali411 Please review.

kshitij12345 committed:
acknowledge complex dtype support in torch.vander

anjali411 commented on test/test_torch.py:
```diff
@@ -7535,6 +7535,18 @@ def test_flip(self, device):
         self.assertEqual(torch.tensor([8, 7, 6, 5, 4, 3, 2, 1]).view(2, 2, 2), data.flip(0, 1, 2))
 
+        # Complex Dtypes
+        def test_complex_flip(dtype):
```
I think we can just update the existing test [here](https://github.com/pytorch/pytorch/blob/08304ccccc11e1a2152d52a8c74b5e88e71c2e6e/test/test_torch.py#L7529)

```
@dtypes(torch.float, torch.cfloat)
def test_flip(self, device, dtype):
        data = torch.randn(8, dtype=dtype, device=device).view(2, 2, 2)
        ...
```

kshitij12345 replied:
The added test is actually a nested function in the mentioned test itself. I'll rename it to `_test_complex_flip` so that it doesn't seem to be an independent test from its name.

kshitij12345 replied:
My bad didn't read the comment carefully. Will do as suggested.

kshitij12345 replied:
Actually using `dtype`, would mean a lot of changes to the test. Every place where `torch.tensor` is initialized to list will have to add `dtype`. Also `arange` doesn't support `complex` dtypes so will have to add `.to(dtype)`
https://github.com/pytorch/pytorch/blob/08304ccccc11e1a2152d52a8c74b5e88e71c2e6e/test/test_torch.py#L7529-L7584

Instead we can have a simple nested function in the test.
```python

        # Complex Dtypes
        def _test_complex_flip(dtype):
            complex_data = torch.tensor([1, 2, 3, 4, 5, 6, 7, 8], device=device, dtype=dtype).view(2, 2, 2)
            self.assertEqual(torch.tensor([5, 6, 7, 8, 1, 2, 3, 4], dtype=dtype).view(2, 2, 2), complex_data.flip(0))
            self.assertEqual(torch.tensor([3, 4, 1, 2, 7, 8, 5, 6], dtype=dtype).view(2, 2, 2), complex_data.flip(1))
            self.assertEqual(torch.tensor([2, 1, 4, 3, 6, 5, 8, 7], dtype=dtype).view(2, 2, 2), complex_data.flip(2))
            self.assertEqual(torch.tensor([7, 8, 5, 6, 3, 4, 1, 2], dtype=dtype).view(2, 2, 2), complex_data.flip(0, 1))
            self.assertEqual(torch.tensor([8, 7, 6, 5, 4, 3, 2, 1], dtype=dtype).view(2, 2, 2), complex_data.flip(0, 1, 2))

        _test_complex_flip(torch.complex64)
        _test_complex_flip(torch.complex128)
```

Do let me know what you feel is a better approach.

anjali411 replied:
hmm I guess then we can just add separate tests in `TestTorchDeviceType`:
```
@dtypes(torch.cfloat, torch.cdouble)
def test_complex_flip(dtype):
     complex_data = torch.randn(8, dtype=dtype, device=device).view(2, 2, 2)
     ...
```
and similar thing for rot90

anjali411 commented on test/test_torch.py:
```diff
@@ -7590,6 +7602,17 @@ def test_rot90(self, device):
         self.assertEqual(torch.tensor([4, 3, 2, 1]).view(2, 2), data.rot90(2, [0, 1]))
         self.assertEqual(torch.tensor([3, 1, 4, 2]).view(2, 2), data.rot90(3, [0, 1]))
 
+        def test_complex_rot90(dtype):
```
same as above. we can just update this [test](https://github.com/pytorch/pytorch/blob/08304ccccc11e1a2152d52a8c74b5e88e71c2e6e/test/test_torch.py#L7586)

kshitij12345 replied:
Similar to above will rename this nested function to `_test_complex_rot90`

anjali411 commented on test/test_torch.py:
```diff
@@ -11226,11 +11249,6 @@ def test_vander(self, device):
             torch.vander(torch.stack((x, x)))
 
-        # This passes on the xla backend
-        if device != 'xla':
```
what is this change for?

kshitij12345 replied:
Similar to `rot90`,`vander` also uses `flip` internally.

https://github.com/pytorch/pytorch/blob/25e6129c5258eb966c8e843ab43e7b4ebaa1b8de/aten/src/ATen/native/TensorFactories.cpp#L920-L946

The removed lines, used to assert that passing a complex dtype will raise a runtime error on `CPU` and `CUDA` backends (not sure why it didn't on `xla`)

However with this PR that won't be the case and thus have removed it.

mruberry replied:
If you're going to make Vander work for complex values you need to add a test for Vander's behavior on complex values, too. You could add a case, for example, to test_vander_types (below), which compares with NumPy as I'm recommending the two tests you're adding do, too. Your case would look something like:

```
if dtype is complex:
  x = [[1 + 1j, 2 - 2j...], ...]]
```

kshitij12345 replied:
`vander` uses `at::cumprod` which doesn't support complex dtypes on `cuda` .

https://github.com/pytorch/pytorch/blob/ff9a809ccd5bd61f18dfaa04d41b681e600b3a02/aten/src/ATen/native/TensorFactories.cpp#L939-L942

So have added test for only CPU.

mruberry replied:
That's good but then the check should be `if self.device_type == `'cuda'` and you should add a comment clarifying why the CUDA device type raises.

kshitij12345 committed:
rename the nested test_complex_* functions

anjali411 commented on aten/src/ATen/native/TensorTransformations.cpp:
```diff
@@ -61,7 +61,7 @@ Tensor flip_cpu(const Tensor& self, IntArrayRef dims) {
   }
 
-  AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Bool, in_tensor.scalar_type(), "flip_cpu", [&] {
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(at::ScalarType::Bool, in_tensor.scalar_type(), "flip_cpu", [&] {
```
let's try and use `AT_DISPATCH_ALL_TYPES_AND_C10_COMPLEX_AND`

anjali411 commented on aten/src/ATen/native/cuda/TensorTransformations.cu:
```diff
@@ -87,7 +87,7 @@ Tensor flip_cuda(const Tensor& self, IntArrayRef dims) {
   // use kernel_pointwise_flip_apply2 only when to-flip dim is the 1st or last dim, where collapseDims can reduce the amount of work
   if (flip_dims_size == 1 && in_tensor.is_contiguous() && (flip_dims[0] == 0 || flip_dims[0] == total_dims - 1)) {
-    AT_DISPATCH_ALL_TYPES_AND2(at::ScalarType::Half, at::ScalarType::Bool, in_tensor.scalar_type(), "flip_cuda", [&] {
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(at::ScalarType::Half, at::ScalarType::Bool, in_tensor.scalar_type(), "flip_cuda", [&] {
```
same as above. `AT_DISPATCH_ALL_TYPES_AND_C10_COMPLEX_AND2`

mruberry commented on test/test_torch.py:
```diff
@@ -7535,6 +7535,18 @@ def test_flip(self, device):
 
+        # Complex Dtypes
+        def test_complex_flip(dtype):
+            complex_data = torch.tensor([1, 2, 3, 4, 5, 6, 7, 8], device=device, dtype=dtype).view(2, 2, 2)
```
These complex values aren't very interesting since they all have zero imaginary part. Instead, I suggest you generate random tensors of various sizes and compare the actual torch result of flipping to the expected NumPy result.

anjali411 replied:
agreed

mruberry commented on test/test_torch.py:
```diff
@@ -7590,6 +7602,17 @@ def test_rot90(self, device):
         self.assertEqual(torch.tensor([3, 1, 4, 2]).view(2, 2), data.rot90(3, [0, 1]))
 
+        def test_complex_rot90(dtype):
+            data = torch.arange(1, 5, device=device).view(2, 2)
```
Again, these tensors don't have the most interesting complex values. Comparing with NumPy will let you populate random values.

anjali411 commented:
@kshitij12345 any updates?

kshitij12345 commented:
> @kshitij12345 any updates?

Hi,
Apologies for the delayed response.
Will address the comments before Saturday.

kshitij12345 committed:
address comments

* update tests to compare with numpy on random data.
* use c10_complex dispatch

kshitij12345 committed:
make flak8 happy :)

kshitij12345 committed:
add test_vander_complex_types CPU

mruberry commented on test/test_torch.py:
```diff
@@ -7590,6 +7609,21 @@ def test_flip(self, device):
 
+    @dtypes(torch.cfloat, torch.cdouble)
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_complex_flip(self, device, dtype):
```
Is there an existing flip tests and/or would it make sense to extend this one to other dtypes, too?

anjali411 replied:
https://github.com/pytorch/pytorch/pull/37826#discussion_r420353275

kshitij12345 replied:
@anjali411 Thanks for adding the reference. Just to expand a bit more on it. The existing tests don't compare against `numpy` but against an expected result. Also they don't explicitly use the `dtype` and thus multiple changes would be needed to support taking the `dtype` argument.

mruberry replied:
OK.

mruberry commented on test/test_torch.py:
```diff
@@ -5977,6 +5978,24 @@ def _np_compare(self, fn_name, vals, device, dtype):
+        torch_fn, np_fn = fns
+
+        a = np.array(vals, dtype=torch_to_numpy_dtype_dict[dtype])
+        np_result = torch.from_numpy(np_fn(a).copy())
```
Why do you need to make a `.copy()` of the result?

kshitij12345 replied:
As flip manipulates the strides and `numpy` supports negative strides, 
without the `.copy()` we get, 
```
ValueError: At least one stride in the given numpy array is negative, and tensors with negative strides are not currently supported. (You can probably work around this by making a copy of your array  with array.copy()
```

mruberry replied:
Would you add a comment explaining that?

kshitij12345 replied:
Done.

mruberry commented on test/test_torch.py:
```diff
@@ -7623,6 +7657,16 @@ def test_rot90(self, device):
 
+    @dtypes(torch.cfloat, torch.cdouble)
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_complex_rot90(self, device, dtype):
```
Same question as above: is there an existing test for rot90 and/or should this test be expanded with additional dtypes?

mruberry commented on test/test_torch.py:
```diff
@@ -11278,6 +11322,30 @@ def test_vander_types(self, device, dtype):
+    @unittest.skipIf(not TEST_NUMPY, 'NumPy not found')
+    @onlyCPU
+    @dtypes(torch.cfloat, torch.cdouble)
+    def test_vander_complex_types(self, device, dtype):
```
You don't need a separate test case, I think? What if you added another case to the existing vander test?

```
elif dtype is torch.cfloat or dtype is torch.cdouble:
  x = [[1 + 1j,...
```

kshitij12345 replied:
The reason I seperated it was because, due to `legacy::cumprod`, `cuda` backend doesn't support `complex dtypes`. The existing case checks with `cpu` and `cuda` both.

So instead of having an `if self.device == 'cuda' and dtype == complexdtype`, made a test with `onlyCPU` decorator which I feel is more explicit in saying the same. 

Once `cumprod` supports `cfloat and cdouble` on CUDA, I believe that is when we should merge the test.

mruberry replied:
Sounds good. Would you add a comment before the new test clarifying that?

mruberry replied:
Actually - one caveat. In the new test, instead of only running it on CPU, would you run it on both and assert that the CUDA version of the test throws a runtime error?

```
if self.device_type == 'cuda':
  with self.assertRaises(RuntimeError):
    ...
```

that way if someone makes cumprod work for complex values on CUDA they'll know to update this test.

kshitij12345 replied:
Makes sense. Will do that.

mruberry reviewed (changes requested):
Looking pretty good but a few questions/suggestions.

kshitij12345 committed:
address comment

* Execute test_vander_complex_types on CPU and CUDA.

kshitij12345 commented:
@pytorchbot retest this please!

kshitij12345 commented:
@mruberry @anjali411 PTAL:)

mruberry commented on test/test_torch.py:
```diff
@@ -11323,6 +11367,38 @@ def test_vander_types(self, device, dtype):
+
+            # Currently `CUDA` backend is not supported as
+            # at::legacy::cumprod does not support complex dtypes.
+            if 'cuda' in device:
```
if self.device_type == 'cuda':

kshitij12345 replied:
Done.

mruberry reviewed (approved):
Nice work. I made a few minor suggestions but overall looks good. Thanks for adding the additional testing.

@anjali411 Want to take a look?

kshitij12345 committed:
address comments

* add comment regarding use of `.copy` on ndarray.
* use self.device_type over device.

anjali411 commented:
great work! can you also add tests for autograd by including `flip` and `rot90` in the complex white list [here](https://github.com/pytorch/pytorch/blob/8292742ba020fcff90f14418c18741ebf606103b/test/test_autograd.py#L4196)

anjali411 commented:
Also I think the tests for `flip` and `rot90` in `tensor_op_tests` (in `test_torch.py`) can be extended to complex as well, and we prob don't need to add any new tests?

cc. @mruberry

mruberry commented:
@anjali411 and I had a chance to talk offline and the new tests are great. We could add a little more coverage by extending the dtypes in some other test suites, but that's OK to leave for a follow-up, too.

kshitij12345 committed:
add test for complex autograd

kshitij12345 committed:
add complex test in tensor_op_tests

kshitij12345 committed:
make flake8 happy:)

kshitij12345 committed:
remove redefinition of _complex_types

kshitij12345 commented:
@mruberry @anjali411 

* Have added to the list in `autograd`.
* And also added complex_dtypes in `tensor_op_tests`, but I have a doubt/question explained below.

I guess there is some issue with 

test_torch.py : Line 17717
```
_signed_types_no_half = [
    torch.float, torch.double,
    torch.int8, torch.short, torch.int, torch.long
]

_cpu_types = []

_unsigned_types = [torch.uint8]
```
Here `_cpu_types` is an empty list.

test_torch.py : Line 18234:
```python

    # Constructs decorator list and applies decorators
    if decorators is None:
        decorators = [dtypes(*dtype_list)]
    else:
        decorators = decorators + [dtypes(*dtype_list)]
    decorators = decorators + [dtypesIfCPU(*dtype_cpu_list)]

    for dec in decorators:
        fn = dec(fn)

    setattr(cls, test_name, fn)
```

Which actually leads to skipping the CPU tests for all types.

For `flip` and `rot90`, I guess `_cpu_types` is used to actually make `inplace=False`.
```
# Each tests is defined in tensor_op_tests as a tuple of:
# - op name (string)
# - (sub)test name (string)
# - tensor constructor, takes dtype and device and constructs the tensor to run the op on
# - arg constructor, takes dtype and device and constructs op arguments
# - torch.half precision (=1e-5)
# - torch.bfloat16 precision (=1e-5)
# - precision (=1e-5), precision to use for all other dtypes
# - dtype_list (=_types), a list of torch dtypes to test the op(s) with
# - cpu_dtype_list (=[]), a list of torch dtypes to test the op(s) on cpu
# - make_inplace_variant (=True), if true the inplace version of the op (op_) is also tested
# - decorators (=[]), a list of decorators to apply to the test
```


So to avoid the skipping, I have used 
```
('flip', 'd0', _small_3d, lambda t, d: [0], 1e-5, 1e-5, 1e-5, _types + _complex_types, \
        _types_no_half + _complex_types, False),
```
instead of 
```
('flip', 'd0', _small_3d, lambda t, d: [0], 1e-5, 1e-5, 1e-5, _types + _complex_types, \
        _cpu_types + _complex_types, False),
```
Below are the run tests with respective list:
[with_cpu_types.txt](https://github.com/pytorch/pytorch/files/4650812/with_cpu_types.txt)
[with_types_no_half.txt](https://github.com/pytorch/pytorch/files/4650813/with_types_no_half.txt)

**Note** : I think `test_torch` should be split or refactored, looks like github doesn't render file of more than 11387 lines.

https://github.com/pytorch/pytorch/blob/dc2e0be791a26acaf64c6e322f0a85a7c32c68bc/test/test_torch.py#L11387

kshitij12345 committed:
use _cpu_dtypes

tensor_op_tests compare device result with cpu result.
So there is no point in having dtypes for CPU to compare
against itself.

anjali411 reviewed (approved):
LGTM thanks :)

anjali411 commented:
@kshitij12345 can you also add autograd tests for `flip` and `rot90` (for complex dtypes) by adding them to the white list in `test_autograd.py`

kshitij12345 commented:
@anjali411
Have already added them in the whitelist.

anjali411 commented:
> @anjali411
> Have already added them in the whitelist.

sorry didn't realize! thanks :D

kshitij12345 commented:
@mruberry 

Sorry was retriggering the CI. The PR can be imported. 
Thanks.

mruberry commented:
Darn, sorry @kshitij12345, looks like a merge conflict snuck in, would you merge this in? Also I'm going to add one nit comment, too, since we have to update the PR, anyway.

mruberry commented on test/test_torch.py:
```diff
@@ -11548,7 +11596,7 @@ def test_vander(self, device):
 
         # This passes on the xla backend
-        if device != 'xla':
+        if device != 'xla' and device != 'cpu':
```
`self.device_type != 'xla' and self.device_type != 'cpu':`

kshitij12345 replied:
Done.

kshitij12345 committed:
address comment

kshitij12345 committed:
update complex dispatch macro

kshitij12345 commented:
@mruberry 
No worries. Done.
Thank You.

@anjali411 
I have reverted to using `AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND{2}` as 
`AT_DISPATCH_ALL_TYPES_AND_C10_COMPLEX_AND2` does not exist (only `AT_DISPATCH_ALL_TYPES_AND_C10_COMPLEX_AND` odes.
 `*_AND2` was removed it seems).

So for consistency have reverted `AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND{2}`.

Thank You.

anjali411 commented:
> @mruberry
> No worries. Done.
> Thank You.
> 
> @anjali411
> I have reverted to using `AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND{2}` as
> `AT_DISPATCH_ALL_TYPES_AND_C10_COMPLEX_AND2` does not exist (only `AT_DISPATCH_ALL_TYPES_AND_C10_COMPLEX_AND` odes.
> `*_AND2` was removed it seems).
> 
> So for consistency have reverted `AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND{2}`.
> 
> Thank You.

@kshitij12345 yeah that sounds fine. we are in the process of migrating from std::complex to c10::complex and those c10 macros are temporary and will be removed soon.

kshitij12345 commented:
@mruberry Gentle ping:)

mruberry commented:
Thanks for the ping, @kshitij12345, we just got back from holiday. I'll take a look today.

mruberry commented:
Sorry for the delayed response, @kshitij12345, just waiting on internal tests.

kshitij12345 commented:
@mruberry 
No worries.
Thank You.

mruberry commented on test/test_torch.py:
```diff
@@ -11692,6 +11740,38 @@ def test_vander_types(self, device, dtype):
+            self.assertEqual(
+                pt_res,
+                torch.from_numpy(np_res),
+                atol=1e-3,
```
You need to add rtol=0

kshitij12345 replied:
Done.

mruberry commented:
Sorry @kshitij12345, one more rebase required. assertEqual now requires specifying rtol when specifying atol for clarity. I made a note in the code where I think this error is coming up.

kshitij12345 committed:
address comment

* add `rtol=0` in assertEqual call.

mruberry commented:
The Vander test failures are real: 

```
May 29 01:19:45 ======================================================================
May 29 01:19:45 ERROR [0.037s]: test_vander_complex_types_cuda_complex128 (__main__.TestTorchDeviceTypeCUDA)
May 29 01:19:45 ----------------------------------------------------------------------
May 29 01:19:45 Traceback (most recent call last):
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 234, in instantiated_test
May 29 01:19:45     result = test(self, device_arg, dtype)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 501, in only_fn
May 29 01:19:45     return fn(self, device, *args, **kwargs)
May 29 01:19:45   File "test_torch.py", line 11931, in test_vander_complex_types
May 29 01:19:45     pt_res = torch.vander(pt_x, increasing=inc) if n is None else torch.vander(pt_x, n, inc)
May 29 01:19:45 RuntimeError: "cumprod_cuda" not implemented for 'ComplexDouble'
May 29 01:19:45 
May 29 01:19:45 ======================================================================
May 29 01:19:45 ERROR [0.038s]: test_vander_complex_types_cuda_complex64 (__main__.TestTorchDeviceTypeCUDA)
May 29 01:19:45 ----------------------------------------------------------------------
May 29 01:19:45 Traceback (most recent call last):
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 234, in instantiated_test
May 29 01:19:45     result = test(self, device_arg, dtype)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 501, in only_fn
May 29 01:19:45     return fn(self, device, *args, **kwargs)
May 29 01:19:45   File "test_torch.py", line 11931, in test_vander_complex_types
May 29 01:19:45     pt_res = torch.vander(pt_x, increasing=inc) if n is None else torch.vander(pt_x, n, inc)
May 29 01:19:45 RuntimeError: "cumprod_cuda" not implemented for 'ComplexFloat'
May 29 01:19:45 
May 29 01:19:45 ======================================================================
May 29 01:19:45 FAIL [0.031s]: test_vander_types_cuda_bool (__main__.TestTorchDeviceTypeCUDA)
May 29 01:19:45 ----------------------------------------------------------------------
May 29 01:19:45 Traceback (most recent call last):
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 234, in instantiated_test
May 29 01:19:45     result = test(self, device_arg, dtype)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 501, in only_fn
May 29 01:19:45     return fn(self, device, *args, **kwargs)
May 29 01:19:45   File "test_torch.py", line 11903, in test_vander_types
May 29 01:19:45     else torch.vander(pt_x, n, inc)
May 29 01:19:45 AssertionError: RuntimeError not raised
May 29 01:19:45 
May 29 01:19:45 ======================================================================
May 29 01:19:45 FAIL [0.031s]: test_vander_types_cuda_float32 (__main__.TestTorchDeviceTypeCUDA)
May 29 01:19:45 ----------------------------------------------------------------------
May 29 01:19:45 Traceback (most recent call last):
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 234, in instantiated_test
May 29 01:19:45     result = test(self, device_arg, dtype)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 501, in only_fn
May 29 01:19:45     return fn(self, device, *args, **kwargs)
May 29 01:19:45   File "test_torch.py", line 11903, in test_vander_types
May 29 01:19:45     else torch.vander(pt_x, n, inc)
May 29 01:19:45 AssertionError: RuntimeError not raised
May 29 01:19:45 
May 29 01:19:45 ======================================================================
May 29 01:19:45 FAIL [0.031s]: test_vander_types_cuda_float64 (__main__.TestTorchDeviceTypeCUDA)
May 29 01:19:45 ----------------------------------------------------------------------
May 29 01:19:45 Traceback (most recent call last):
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 234, in instantiated_test
May 29 01:19:45     result = test(self, device_arg, dtype)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 501, in only_fn
May 29 01:19:45     return fn(self, device, *args, **kwargs)
May 29 01:19:45   File "test_torch.py", line 11903, in test_vander_types
May 29 01:19:45     else torch.vander(pt_x, n, inc)
May 29 01:19:45 AssertionError: RuntimeError not raised
May 29 01:19:45 
May 29 01:19:45 ======================================================================
May 29 01:19:45 FAIL [0.032s]: test_vander_types_cuda_int16 (__main__.TestTorchDeviceTypeCUDA)
May 29 01:19:45 ----------------------------------------------------------------------
May 29 01:19:45 Traceback (most recent call last):
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 234, in instantiated_test
May 29 01:19:45     result = test(self, device_arg, dtype)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 501, in only_fn
May 29 01:19:45     return fn(self, device, *args, **kwargs)
May 29 01:19:45   File "test_torch.py", line 11903, in test_vander_types
May 29 01:19:45     else torch.vander(pt_x, n, inc)
May 29 01:19:45 AssertionError: RuntimeError not raised
May 29 01:19:45 
May 29 01:19:45 ======================================================================
May 29 01:19:45 FAIL [0.031s]: test_vander_types_cuda_int32 (__main__.TestTorchDeviceTypeCUDA)
May 29 01:19:45 ----------------------------------------------------------------------
May 29 01:19:45 Traceback (most recent call last):
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 234, in instantiated_test
May 29 01:19:45     result = test(self, device_arg, dtype)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 501, in only_fn
May 29 01:19:45     return fn(self, device, *args, **kwargs)
May 29 01:19:45   File "test_torch.py", line 11903, in test_vander_types
May 29 01:19:45     else torch.vander(pt_x, n, inc)
May 29 01:19:45 AssertionError: RuntimeError not raised
May 29 01:19:45 
May 29 01:19:45 ======================================================================
May 29 01:19:45 FAIL [0.032s]: test_vander_types_cuda_int64 (__main__.TestTorchDeviceTypeCUDA)
May 29 01:19:45 ----------------------------------------------------------------------
May 29 01:19:45 Traceback (most recent call last):
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 234, in instantiated_test
May 29 01:19:45     result = test(self, device_arg, dtype)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 501, in only_fn
May 29 01:19:45     return fn(self, device, *args, **kwargs)
May 29 01:19:45   File "test_torch.py", line 11903, in test_vander_types
May 29 01:19:45     else torch.vander(pt_x, n, inc)
May 29 01:19:45 AssertionError: RuntimeError not raised
May 29 01:19:45 
May 29 01:19:45 ======================================================================
May 29 01:19:45 FAIL [0.031s]: test_vander_types_cuda_int8 (__main__.TestTorchDeviceTypeCUDA)
May 29 01:19:45 ----------------------------------------------------------------------
May 29 01:19:45 Traceback (most recent call last):
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 234, in instantiated_test
May 29 01:19:45     result = test(self, device_arg, dtype)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 501, in only_fn
May 29 01:19:45     return fn(self, device, *args, **kwargs)
May 29 01:19:45   File "test_torch.py", line 11903, in test_vander_types
May 29 01:19:45     else torch.vander(pt_x, n, inc)
May 29 01:19:45 AssertionError: RuntimeError not raised
May 29 01:19:45 
May 29 01:19:45 ======================================================================
May 29 01:19:45 FAIL [0.031s]: test_vander_types_cuda_uint8 (__main__.TestTorchDeviceTypeCUDA)
May 29 01:19:45 ----------------------------------------------------------------------
May 29 01:19:45 Traceback (most recent call last):
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 798, in wrapper
May 29 01:19:45     method(*args, **kwargs)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 234, in instantiated_test
May 29 01:19:45     result = test(self, device_arg, dtype)
May 29 01:19:45   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 501, in only_fn
May 29 01:19:45     return fn(self, device, *args, **kwargs)
May 29 01:19:45   File "test_torch.py", line 11903, in test_vander_types
May 29 01:19:45     else torch.vander(pt_x, n, inc)
May 29 01:19:45 AssertionError: RuntimeError not raised
May 29 01:19:45 
May 29 01:19:45 ----------------------------------------------------------------------
```

mruberry commented on test/test_torch.py:
```diff
@@ -11842,6 +11890,39 @@ def test_vander_types(self, device, dtype):
+
+            # Currently `CUDA` backend is not supported as
+            # at::legacy::cumprod does not support complex dtypes.
+            if self.device_type == 'cuda':
```
I think this check should be in complex_types, not this test?

kshitij12345 replied:
Yup. You are right. I guess it moved around during the rebase. 

Fixed now.

kshitij12345 committed:
move condition to correct test

kshitij12345 commented:
@mruberry 
I think we should wait for #39063 to land, as it will add complex support for `cumsum` which would break the `vander` test (with `Runtime Error not raised`). 

Instead we should rebase with the master once it is in and test the CUDA as well.

kshitij12345 committed:
Merge branch lastest 'master' with complex cumsum support into support/complex/flip-rot90

kshitij12345 committed:
enable cuda test for vander complex

kshitij12345 committed:
merge complex test with the existing test

kshitij12345 committed:
fix stray space

kshitij12345 commented:
@mruberry 
* Have merged with the latest master thus `vander` now supports complex on CUDA.
* Have merged the complex vander test with the existing one.

Do review. Thanks.

mruberry reviewed (approved):
Awesome!

mruberry commented:
> @mruberry Gentle ping

Thanks @kshitij12345, internal tests looked good. I'll start landing now.

mruberry commented:
Diff is landed and tests are now running on OSS Master to validate.

mruberry commented:
Awesome! OSS Master validation tests look good.

kshitij12345 commented:
Great! Thanks.

Will rebase https://github.com/pytorch/pytorch/pull/38599 and add tests.
Will ping you once ready.
